modules/host/upload.go

```diff
@@ -13,53 +16,447 @@ import (
+		return errors.New("couldn't read contract ID: " + err.Error())
+	}
+
+	// remove conn deadline while we wait for lock and rebuild the Merkle tree.
+	err := conn.SetDeadline(time.Time{})
 	if err != nil {
 		return err
 	}
-	defer file.Close()
 
-	// Process requests until 'stop' signal is received.
-	var request modules.DownloadRequest
-	for {
-		if err := encoding.ReadObject(conn, &request, 16); err != nil {
-			return err
-		}
-		// Check for termination signal.
-		// TODO: perform other sanity checks on offset/length?
-		if request.Length == 0 {
-			break
-		}
+	h.mu.RLock()
+	obligation, exists := h.obligationsByID[fcid]
+	h.mu.RUnlock()
+	if !exists {
+		return errors.New("no record of that contract")
+	}
+
+	// need to protect against two simultaneous revisions to the same
+	// contract; this can cause inconsistency and data loss, making storage
+	// proofs impossible
+	//
+	// TODO: DOS vector - the host has locked the obligation even though the
+	// renter has not proven themselves to be the owner of the file contract.
+	obligation.mu.Lock()
+	defer obligation.mu.Unlock()
+
+	// open the file in append mode
+	file, err := os.OpenFile(obligation.Path, os.O_CREATE|os.O_RDWR|os.O_APPEND, 0660)
+	if err != nil {
+		return err
+	}
+
+	// rebuild current Merkle tree
+	tree := crypto.NewTree()
+	err = tree.ReadSegments(file)
+	if err != nil {
+		// Error does not need to be checked when closing the file, already
+		// there have been issues related to the filesystem.
+		_ = file.Close()
+		return err
+	}
+
+	// accept new revisions in a loop. The final good transaction will be
+	// submitted to the blockchain.
+	revisionErr := func() error {
+		for {
+			// allow 5 minutes between revisions
+			err := conn.SetDeadline(time.Now().Add(5 * time.Minute))
+			if err != nil {
+				return err
+			}
+
+			// read proposed revision
+			var revTxn types.Transaction
+			if err = encoding.ReadObject(conn, &revTxn, types.BlockSizeLimit); err != nil {
+				return errors.New("couldn't read revision: " + err.Error())
+			}
+			// an empty transaction indicates completion
+			if revTxn.ID() == (types.Transaction{}).ID() {
+				return nil
+			}
+
+			// allow 5 minutes for each revision
+			err = conn.SetDeadline(time.Now().Add(5 * time.Minute))
+			if err != nil {
+				return err
+			}
+
+			// check revision against original file contract
+			h.mu.RLock()
+			err = h.considerRevision(revTxn, obligation)
+			h.mu.RUnlock()
+			if err != nil {
+				// There is nothing that can be done if there is an error while
+				// writing to a connection.
+				_ = encoding.WriteObject(conn, err.Error())
+				continue // don't terminate loop; subsequent revisions may be okay
+			}
+
+			// indicate acceptance
+			if err := encoding.WriteObject(conn, modules.AcceptResponse); err != nil {
+				return errors.New("couldn't write acceptance: " + err.Error())
+			}
+
+			// read piece
+			// TODO: simultaneously read into tree and file
```

Comment:
this TODO is probably a bad idea. If something goes wrong you'd need to truncate the file. However, it may be necessary if revisions get larger than a few dozen MB.